client/services/cardKeyService.js

```diff
@@ -74,64 +74,116 @@ module.exports = [
+              // Debounce
+              if (data !== lastData || Date.now() - lastTime > 2000) {
+                // data = card id
+                cb(data);
+                lastTime = Date.now();
+                lastData = data;
+              }
             }
-          }
-        };
+          };
+          const readResult = async () => {
+            const message = [];
+            let finished = false;
+            // Keep reading bytes until the "end" byte is sent
+            // The "end" byte is 0xbb
+            while (!finished) {
+              const { value } = await $rootScope.serialReader.read();
+              for (let i = 0; i < value.length; i++) {
+                // First byte in a message should be 170, otherwise ignore and keep on going
+                if (message.length === 0 && value[i] !== 170) {
+                  continue;
+                }
+                // Second byte in a message should be 255, otherwise discard and keep on going
+                if (message.length === 1 && value[i] !== 255) {
+                  // If value is 170, treat it as the first value, and keep on. Otherwise discard
+                  if (value[i] !== 170) {
+                    message.length = 0;
+                  }
+                  continue;
+                }
 
-        const readResult = async () => {
-          const message = [];
-          let finished = false;
-          // Keep reading bytes until the "end" byte is sent
-          // The "end" byte is 0xbb
-          while (!finished) {
-            const { value } = await $rootScope.serialReader.read();
-            for (let i = 0; i < value.length; i++) {
-              if (value[i] == 0xbb) {
-                finished = true;
-                break;
+                if (message.length > 3 && message.length >= message[2] + 4) {
+                  finished = true;
+                  break;
+                }
+                message.push(value[i]);
               }
-              message.push(value[i]);
             }
-          }
-          onComplete(message);
-        };
+            onComplete(message);
+          };
 
-        // Constantly send the readCardCommand and read the result.
-        // If there is no card, the result will be an error status,
-        // which is handled in the onComplete function
-        $rootScope.serialInterval = setInterval(() => {
-          $rootScope.serialWriter.write(readCardCommand);
-          readResult();
-        }, 500);
+          // Constantly send the readCardCommand and read the result.
+          // If there is no card, the result will be an error status,
+          // which is handled in the onComplete function
+          const runPoll = async () => {
+            try {
+              $rootScope.serialDevice.writer.write(readCardCommand);
+              await readResult();
+            } catch (e) {
+              /* eslint no-console: 0 */
+              console.err('Error doing card stuff', e);
```

Comment:
⛔ 
![image](https://user-images.githubusercontent.com/23152018/107061381-6a5cde00-67d8-11eb-83c9-4145b2860250.png)

Reply:
:shipit:

Reply:
How did you get those errors?